include/aspect/assemblers.h

```diff
@@ -0,0 +1,106 @@
+/*
+  Copyright (C) 2016 by the authors of the ASPECT code.
```

Comment:
I think the idea here is to separate the concrete implementation of all the small assemblers (the function `local_assemble_...`) from the global tasks of assembling all equations. `assembly.h` and `assembly.cc` have become huge and completely confusing over time (more than 3000 lines in assembly.cc on master). 
The new structure in this PR is certainly not perfect either, but at least it separates the global stuff from the assembly of the individual terms.

This also relates to your comment further down about complexity of the assembler signals. I think (and we have discussed this with Timo and Juliane in Clemson) the signals have served their purpose, but in practice we do nothing with them that would not be possible with a vector of pointers to assembler objects. This could have exactly the same plugin structure we use everywhere else in the code, and we could remove the  the only signal we would keep would be the `set_assemblers` signal that allows plugins to modify the vector of pointers to `Assemblers::Interface` objects. This structure should be equally fast, and much easier to understand for users, because it follows the same design pattern as the rest of the code. This would obviously be a larger change that we should not mix with the current PR, but the separation into separate files might be a first step into that direction.
Would you agree with that general direction?